gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFSInputStream.java

```diff
@@ -106,60 +111,82 @@ public synchronized int read() throws IOException {
 
   @Override
   public synchronized int read(@Nonnull byte[] buf, int offset, int length) throws IOException {
-    checkNotClosed();
-    streamStatistics.readOperationStarted(getPos(), length);
-    checkNotNull(buf, "buf must not be null");
-    if (offset < 0 || length < 0 || length > buf.length - offset) {
-      throw new IndexOutOfBoundsException();
-    }
-    int response = 0;
-    try {
-      // TODO(user): Wrap this in a while-loop if we ever introduce a non-blocking mode for the
-      // underlying channel.
-      int numRead = channel.read(ByteBuffer.wrap(buf, offset, length));
-      if (numRead > 0) {
-        // -1 means we actually read 0 bytes, but requested at least one byte.
-        totalBytesRead += numRead;
-        statistics.incrementBytesRead(numRead);
-        statistics.incrementReadOps(1);
-      }
-      response = numRead;
-    } catch (IOException e) {
-      streamStatistics.readException();
-    }
-    streamStatistics.bytesRead(max(response, 0));
-    streamStatistics.readOperationCompleted(length, max(response, 0));
-    return response;
+    return trackDuration(
```

Comment:
It would be much cleaner if we can use AOP and annotation to capture the "duration" of each API.

Reply:
Using the pattern from hadoop. Adding AOP will be more work and I do no think it is worth the effort.